### PR TITLE
Account for more node types in handling of except block homonyms with comprehensions

### LINKED ISSUE #6069: False positive `used-before-assignment` for homonym between subscript in comprehension and assignment in except block
False positive for `used-before-assignment` from subscript notation in comprehension followed by definition of that same variable later in an except block.

```python
# test.py
def select_dict(data, keys):
    try:
        d = {key: data[key] for key in keys}
    except KeyError as e:
        key, *_ = e.args
        raise Exception(f"{key} not found")

    return d
```

```console
$ pip install -U pylint==2.13.4 -q && pylint test.py
************* Module test
test.py:3:23: E0601: Using variable 'key' before assignment (used-before-assignment)

------------------------------------------------------------------
Your code has been rated at 2.86/10 (previous run: 2.86/10, +0.00)
```

_Originally posted by @skshetry in https://github.com/PyCQA/pylint/issues/6035#issuecomment-1084339275_

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Several nodes types didn't meet the allow list for homynym handling b/w except blocks and comprehensions.

Closes #6069

### CODE REVIEW

**Comment by jacobtylerwalls** [comment]
Close, but couldn't avoid breaking this test. Back to whackamole.

https://github.com/PyCQA/pylint/blob/cb9dc791a0e83f90391cc234b03cb5e4769400b3/tests/functional/u/unused/unused_argument.py#L43-L48

**Comment by jacobtylerwalls** [comment]
Ah, okay, I do have another diff that works. But it's not battle-tested enough for a hotfix timeline. I can follow up for 2.14.

File: pylint/checkers/variables.py [hunk 1]
```diff
@@ -2483,24 +2505,25 @@ def _has_homonym_in_comprehension_test(node: nodes.Name) -> bool:
         """Return True if `node`'s frame contains a comprehension employing an
         identical name in a test.
```
Comment (jacobtylerwalls) [inline pylint/checkers/variables.py:2484]: Todo: remove all references to "test"

**Comment by jacobtylerwalls** [comment]
Hi @skshetry, I looked for an approach that didn't involve special-casing each type of `NodeNG` that might appear (subscript, etc.). Could I ask you to test this PR against your project? That would give me more confidence going into a quick patch release timeline. Thanks.

**Comment by jacobtylerwalls** [comment]
anyone is welcome to review, just hitting the button to dismiss the now moot review!

**Review by Pierre-Sassoulas — COMMENTED** [review]
I don't see any problem here, but I'm really out of touch with used-before-assignment at this point 😄 I'm wondering if astroid could help with multiple imbricated scopes in a generic "simpler" way. How is it done in other linter ? Is it supposed to be *that* hard ? Is there an issue with the representation of the AST ? Maybe a state of the art could help ?

File: pylint/checkers/variables.py [hunk 2]
```diff
@@ -1282,8 +1282,31 @@ def leave_functiondef(self, node: nodes.FunctionDef) -> None:
 
         global_names = _flattened_scope_names(node.nodes_of_class(nodes.Global))
         nonlocal_names = _flattened_scope_names(node.nodes_of_class(nodes.Nonlocal))
+        comprehension_target_names = []
+
+        def find_assigned_names_recursive(
```
Comment (DanielNoord) [inline pylint/checkers/variables.py:1287]: I always mess up my declaration order, but doesn't this get redefined for every `FunctionDef` we visit? That might become performance impacting?

**Review by DanielNoord — APPROVED** [review]
I like this approach! Well done figuring it out!

File: pylint/checkers/variables.py (hunk 2 above)
Comment (jacobtylerwalls) [reply pylint/checkers/variables.py:1287]: Yeah I can take a look at pulling this out into a private instance method.

Comment (jacobtylerwalls) [reply pylint/checkers/variables.py:1287]: It was tested. But I'm adding another case with more nesting.

> I always mess up my declaration order, but doesn't this get redefined for every FunctionDef we visit? That might become performance impacting?

Now that I'm looking closely, I guess I don't see a performance impact. Do nested functions really have a worse impact over nested loops and such?

**Comment by jacobtylerwalls** [comment]
> Is it supposed to be that hard ?

Thanks, you inspired me to delete more code in e35d1da.

File: pylint/checkers/variables.py (hunk 2 above)
Comment (DanielNoord) [reply pylint/checkers/variables.py:1287]: I really don't know 😄 

I'm not sure if `find_assigned_names_recursive` is stored in memory or if it is re-declared every time `leave_functiondef` gets called. The later sounds as if that could be problematic.

File: pylint/checkers/variables.py [hunk 3]
```diff
@@ -2479,28 +2501,14 @@ def _has_homonym_in_upper_function_scope(
...
-            node, nodes.Comprehension
-        )
-        return (
-            closest_comprehension is not None
-            and node.frame(future=True).parent_of(closest_comprehension)
-            and any(
-                test is node or test.parent_of(node)
-                for test in closest_comprehension.ifs
-            )
+    def _comprehension_under_frame(node: nodes.Name) -> bool:
```
[the remark this answers is not in the document]
Comment (jacobtylerwalls) [reply pylint/checkers/variables.py]: I guess I don't feel strongly about making it a util. I don't envision many uses for it.

[the remark this answers is not in the document]
Comment (DanielNoord) [reply pylint/checkers/variables.py]: Fine with me!

File: pylint/checkers/variables.py [hunk 4]
```diff
@@ -2264,7 +2279,13 @@ def _loopvar_name(self, node: astroid.Name) -> None:
...
     def _check_is_unused(
-        self, name, node, stmt, global_names, nonlocal_names: Iterable[str]
+        self,
+        name,
+        node,
+        stmt,
+        global_names,
+        nonlocal_names: Iterable[str],
+        comprehension_target_names: List[str],
     ):
```
Comment (DanielNoord) [inline pylint/checkers/variables.py:2289]: This is missing return type annotation.

**Review by DanielNoord — APPROVED** [review]
Only the return type needs to be added 😄

File: pylint/checkers/variables.py (hunk 4 above)
Comment (Pierre-Sassoulas) [reply pylint/checkers/variables.py:2289]:
```suggestion
    ) -> None:
```

**Pull request merged**